### PR TITLE
fix: improve error handling on MLS commits - part 1 AR-2286

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Improve MLS commit error handling to follow the strategies outlined her https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/614236365/Use+case+send+a+commit+message+MLS

### Dependencies

Needs releases with:

- [ ] https://github.com/wireapp/kalium/pull/861

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

### Notes

I plan to improve the error handling for `establishMLSGroup()` and `requestToJoinGroup()` as well but that they require more re-factoring which I want to in a separate PR.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt [hunk 1]
```diff
@@ -145,6 +155,11 @@ class MLSConversationDataSource(
         }
 
     override suspend fun updateKeyingMaterial(groupID: GroupID): Either<CoreFailure, Unit> =
+        executeOperation(groupID) {
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:158]: isn't better instead of having two functions, wrap the body of the function with executeOperation wrapper? like other wrappers. 
I know that we need to keep the main functions unwrapped to use them in other places and wrap the whole new function[my assumption] but still we can improve this of having duplicated functions.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt [hunk 2]
```diff
@@ -277,4 +304,65 @@ class MLSConversationDataSource(
         conversationDAO.getAllMembers(conversationID).first().map { idMapper.fromDaoModel(it.user) }
     }
 
+    private suspend fun executeOperation(groupID: GroupID, operation: suspend () -> Either<CoreFailure, Unit>) =
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:307]: I think we can extract this from the repo and put it in a util class.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt [hunk 3]
```diff
@@ -277,4 +304,65 @@ class MLSConversationDataSource(
...
+        operation()
+            .flatMapLeft {
+                handleCommitFailure(it, groupID, operation)
+            }
+
+    private suspend fun handleCommitFailure(
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:313]: a suggestion for this part of the code:
I think this part would grow more and more in the future, and it's not clean IMO.
My suggestion: to have this functions and the retry mechanism inside the Failure sealed classes, and then here we only call the retry once for all! based on the sealed type the correct function will be invoked.
[if this is not clear, we can have a call to discuss it]

Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:313]: Let's have a call

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt [hunk 4]
```diff
@@ -277,4 +316,52 @@ class MLSConversationDataSource(
         conversationDAO.getAllMembers(conversationID).first().map { idMapper.fromDaoModel(it.user) }
     }
 
+    private suspend fun retryOnCommitFailure(groupID: GroupID, operation: suspend () -> Either<CoreFailure, Unit>) =
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:319]: since it's not always a retry, can we rename it to something like wrapStorageRequest, e.g. wrapMLSCommitRequest?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt [hunk 5]
```diff
@@ -277,4 +316,52 @@ class MLSConversationDataSource(
...
+                handleCommitFailure(it, groupID, operation)
+            }
+
+    private suspend fun handleCommitFailure(
+        failure: CoreFailure,
+        groupID: GroupID,
+        retryOperation: suspend () -> Either<CoreFailure, Unit>
+    ): Either<CoreFailure, Unit> {
+        return when (failure.getStrategy()) {
+            CommitStrategy.KEEP_AND_RETRY -> keepCommitAndRetry(groupID)
```
Comment (mchenani) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:331]: do we have any documentation about when we should retry, abort or discard and retry? I'm curious about future operation that how can we figure out what strategy we should pick for them!

**Review by mchenani — APPROVED** [review]
Clean and needed! just one comment and one question! 🚀

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt (hunk 4 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:319]: It's always a retry if does something otherwise it passes the error on, maybe it would be more accurate to name it `retryOnCommitFailureIfPossible` but I find it too wordy. 

IMO this is not the same as `wrapStorageRequest ` we aren't catching exceptions putting them in a `Either<CoreFailure, Unit>`.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt (hunk 5 above)
Comment (typfel) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt:331]: I'm implementing this https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/614236365/Use+case+send+a+commit+message+MLS
